### DISCUSSION TITLE
Broken links ("Warning: cURL error 7: Failed to connect")

### DISCUSSION THREAD

**Discussion by nunesgh** [nextcloud/bookmarks#1714] [Q&A] [answered]
Hi,

I have recently installed the Bookmarks app and I am still testing it. I was trying to understand exactly what was expected from the screenshot feature --whether an image file would be created in the Files app or something else-- and in the process I have added and deleted the same [link](https://nextcloud.com/blog/decentralize-the-cloud-with-fast-deployment-and-easy-backup/) around half a dozen times.

The last time I added the link back to the Bookmarks app, it was not properly fetched, though, being added to the Broken links section. Also, the following error appeared on the Logging app:
`[bookmarks] Warning: cURL error 7: Failed to connect to nextcloud.com port 443: Connection timed out (see https://curl.haxx.se/libcurl/c/libcurl-errors.html) for https://nextcloud.com/blog/decentralize-the-cloud-with-fast-deployment-and-easy-backup/`

What happens now? Is the Bookmarks app expected to retry fetching that link in the future? If so, how long after the last unsuccessful attempt?

How exactly does the Bookmarks app behave when it comes to checking for broken links?

Thank you for your attention.

**Comment by marcelklehr** [comment, marked as answer]
Hi @nunesgh!

Welcome to the community and thank you for the great question(s)!

> What happens now? Is the Bookmarks app expected to retry fetching that link in the future? If so, how long after the last unsuccessful attempt?

The App has a background job, that regularly crawls all links and will revisit a link after [28 days](https://github.com/nextcloud/bookmarks/blob/master/lib/Service/BookmarkPreviewer.php#L24).

> How exactly does the Bookmarks app behave when it comes to checking for broken links?

The app will consider a link as broken if [one of the following is the case](https://github.com/nextcloud/bookmarks/blob/master/lib/Service/CrawlService.php#L86-L102):

 * A GET request to the URL returns a 404 status code
 * A GET request to the URL takes longer than 10 seconds to establish the connection
 * A GET request to the URL takes longer than 10 seconds until the response is received
 * A GET request to the URL takes longer than 10 seconds to read the response body
 * A GET request to the URL fails e.g. due to Network or DNS problems

**Reply by nunesgh** [reply]
Thank you for your detailed reply, @marcelklehr!

Is there a reason for the hard-coded 28 days?

**Reply by marcelklehr** [reply]
It's essentially a trade-off between compute and having up-to-date data. 28 days seems reasonable to me.
